src/coreclr/jit/gentree.cpp

```diff
@@ -19740,6 +19740,84 @@ GenTree* Compiler::gtNewSimdBinOpNode(
 
             switch (simdBaseType)
             {
+                case TYP_BYTE:
+                case TYP_UBYTE:
+                {
+                    assert((simdSize != 64) || IsBaselineVector512IsaSupportedDebugOnly());
+
+                    CorInfoType widenSimdBaseJitType;
+                    NamedIntrinsic convertIntrinsic;
+                    var_types      convertedType;
+                    unsigned       convertedSimdSize;
+
+                    if (simdSize == 32 && IsBaselineVector512IsaSupported())
+                    {
+                        if (simdBaseType == TYP_BYTE)
+                        {
+                            widenSimdBaseJitType = CORINFO_TYPE_SHORT;
+                            convertIntrinsic     = NI_AVX512BW_ConvertToVector512Int16;
+                        }
+                        else
+                        {
+                            widenSimdBaseJitType = CORINFO_TYPE_USHORT;
+                            convertIntrinsic     = NI_AVX512BW_ConvertToVector512UInt16;
+                        }
+
+                        convertedType     = TYP_SIMD64;
+                        convertedSimdSize = 64;
+                    }
+                    else if (simdSize == 16 && compOpportunisticallyDependsOn(InstructionSet_AVX2))
+                    {
+                        widenSimdBaseJitType = simdBaseType == TYP_BYTE ? CORINFO_TYPE_SHORT : CORINFO_TYPE_USHORT;
+                        convertIntrinsic = NI_AVX2_ConvertToVector256Int16;
+                        convertedType     = TYP_SIMD32;
+                        convertedSimdSize = 32;
+                    }
+                    else
+                    {
+                        widenSimdBaseJitType = simdBaseType == TYP_BYTE ? CORINFO_TYPE_SHORT : CORINFO_TYPE_USHORT;
+                        convertIntrinsic  = NI_Illegal;
+                        convertedType     = TYP_UNDEF;
+                        convertedSimdSize = 0;
+                    }
+
+                    if (convertedType != TYP_UNDEF)
+                    {
+                        op1 = gtNewSimdHWIntrinsicNode(convertedType, op1, convertIntrinsic, simdBaseJitType,
+                                                       simdSize);
+
+                        op2 = gtNewSimdHWIntrinsicNode(convertedType, op2, convertIntrinsic, simdBaseJitType,
+                                                       simdSize);
+
+                        op1 = gtNewSimdBinOpNode(GT_MUL, convertedType, op1, op2, widenSimdBaseJitType, convertedSimdSize);
+
+                        GenTree* dup = gtCloneExpr(op1);
```

Comment:
This does not look correct in case `op1` had persistent side effects (e. g. a call).

Reply:
The reason for this is that using `fgMakeMultiUse` generates this codegen
```asm
G_M65351_IG01:  ;; offset=0000H
       vzeroupper 
						;; size=3 bbWeight=1 PerfScore 1.00
G_M65351_IG02:  ;; offset=0003H
       vpmovzxbw ymm0, qword ptr [rdx]
       vpmovzxbw ymm1, qword ptr [r8]
       vpmullw  ymm0, ymm0, ymm1
       vmovups  ymm1, ymmword ptr [reloc @RWD00]
       vpand    ymm2, ymm0, ymm1
       vpand    ymm0, ymm0, ymm1
       vpackuswb ymm0, ymm2, ymm0
       vpermq   ymm0, ymm0, -40
       vmovups  xmmword ptr [rcx], xmm0
       mov      rax, rcx
						;; size=47 bbWeight=1 PerfScore 23.92
G_M65351_IG03:  ;; offset=0032H
       vzeroupper 
       ret      
						;; size=4 bbWeight=1 PerfScore 2.00
RWD00  	dq	00FF00FF00FF00FFh, 00FF00FF00FF00FFh, 00FF00FF00FF00FFh, 00FF00FF00FF00FFh
```
where the `vpand` is performed twice on the same input.

Notice that here `op1` is the result of the multiplication, which at this point should be guaranteed to be an hardware instrinsic, as far as I understand.

Reply:
> Notice that here op1 is the result of the multiplication, which at this point should be guaranteed to be an hardware instrinsic, as far as I understand.

While the node itself is an HWI at this point, the tree (which includes the original `op1`) is not. If that tree contains a store/call/etc, duplicating it is not legal (and duplicating trees like this is perilous in general, as they can be arbitrarily expensive to evaluate); the codegen issue will have to be resolved some other way.

Reply:
I will change this to `fgMakeMultiUse` and investigate the behavior.
What I noticed is that this helper method has only been used with a `GenTree*` [extracted from the stack](https://github.com/dotnet/runtime/blob/87d6bbaa59aeeb9bfdca08a01e12bd87e4f2fbdd/src/coreclr/jit/hwintrinsicxarch.cpp#L2118) until now, and in such contexts does not present that behavior.

Reply:
Fixed code as per review on 68467333f6d8f6bc9789a910dd0028a2325b651c.